Mangage Projects & Files in the Backend

Henotu opened the pull request:
This pull request adds support for managing files and projects in the backend.
The backend provides an API which is documented using an openAPI specification (`backend/src/main/resources/api.yaml`).

Furthermore, the API ist testet using automatic unit tests.

Closes #5
Closes #3

Henotu committed:
Add first draft of specification

Henotu committed:
Add further HTTP-methods to endopoints

Also added the endpoint /file/ to allow for file creation.

Henotu committed:
Introduced Directory as an explicit type

Henotu committed:
Add representation of FileElement

Henotu committed:
Add custom ID generator

The way in which IDs are generated should be seen as a temporary
representation. It is left open if the IDs switch to UUIDs later in
development.

Henotu committed:
Add Repository for File and Directory

Henotu committed:
Add first version of FileController

This version still has some missing implementation (PATCH File) and some
missing functionality (GET/PUT file content).

Henotu committed:
Implement File-patching

The current implementation is not greatly expandable, see the
switch-case statements. This has to be refactored in later commits.

Henotu committed:
Add file-content to API

Henotu committed:
Implement patch method of FileElement

Henotu committed:
Add Project to model and controller

Henotu committed:
Add Tests for FileController

Also fixed bugs that were uncovered during testing.

Henotu committed:
Fix bad request error missing when patching dir.

The method would not throw anything when the user tries to patch extra
content.

Henotu committed:
Add tests for ProjectController and fix bugs

Henotu committed:
Merge branch 'openapi-specification' into 'feature/5-manage-projects-backend'

Henotu committed:
Fix project content not being able to get deleted

Henotu committed:
Update API documentation

Main change is the switch from integer-based IDs to string-based IDs

Henotu committed:
Delete unnecessary test

Henotu committed:
Remove display of lower level contents in API-call

Henotu committed:
Add jdoc and refactor some classes

Henotu committed:
Fix contents-filter in project overview

Jooshua commented on backend/build.gradle:
```diff
@@ -23,6 +23,7 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
 	implementation 'org.springframework.boot:spring-boot-starter-jdbc'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'com.google.code.gson:gson:2.13.1'
```
Why are we using GSON and Jackson? Wouldn't one be sufficient?

Henotu replied:
Fair point. Gson was only during testing, which I now changed to Jackson.

Jooshua replied:
So is Gson still required? Or can we ger rid of that dependency?

Henotu replied:
The preview above is outdated. The dependency was removed in commit 9dad3656589ebef6c1bbfbc3c111652271fdd110

Jooshua commented on backend/src/main/java/edu/kit/quak/files/model/Type.java:
```diff
@@ -0,0 +1,71 @@
+/*
+ * TODO: The information inside this enum should be moved to their respective classes
+ *       to allow for better extensibility
+ */
```
Should this TODO be resolved before merging?

Henotu replied:
Nice catch, I forgot about this TODO when creating the PR.

The refactoring that resulted from Removing the `Type`-Enum introduced some further methods to FileElement and some classes in `edu.kit.quak.files.repository.savers`.
The main benefit of this change is that the `FileController` no longer uses switch-case statements to differenciate between files and directories.

Jooshua commented on backend/src/test/java/edu/kit/quak/QuaKApplicationTests.java:
```diff
@@ -1,13 +1,30 @@
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
 @SpringBootTest
-class QuaKApplicationTests {
+public class QuaKApplicationTests {
```
Should this be an abstract class?

Jooshua replied:
I am also a fan of naming conventions for abstract class (A...) and interfaces (I...), could we do that for the Java part of the project?

Henotu replied:
`QuaKApplicationTests` is now abstract.

> I am also a fan of naming conventions for abstract class (A...) and interfaces (I...), could we do that for the Java part of the project?

Uusing such names would encode the type in the name, which is redundant and in my opinion unnecessary, especially since modern IDEs can easily display the type of a class.
On the other hand, you're the product owner. If you prefer this naming convention, I could implement it.

Jooshua commented on backend/src/main/resources/api.yaml:
```diff
@@ -0,0 +1,390 @@
+          enum: [file, directory]
+
+
+    File:
+      allOf:
+        - $ref: '#/components/schemas/FileElement'
+        - type: object
+          properties:
+            lastAccess:
+              type: integer
+              format: int64
+              description: Unix-Epoch in seconds
+            createdOn:
+              type: integer
+              format: int64
+              description: Unix-Epoch in seconds
```
Do we wan't to distinguish different content types? I think this could be beneficial in the future.

Henotu replied:
I'm not quite sure what you mean by that. Do you want to have the `Content-Type` displayed in a file-object, similar to the `lastAccess` field?

Jooshua commented on backend/src/main/java/edu/kit/quak/files/configuration/DepthFilter.java:
```diff
@@ -0,0 +1,25 @@
+
+    private static final HashSet<String> IGNORE = new HashSet<>(List.of("contents"));
+
+    private final static int MAX_DEPTH = 3;
```
Just curious how this works, as we don't use `MAX_DEPTH` anywhere?

Henotu replied:
This was an artifact from a previous implementation of the filter. It is now removed

Jooshua commented on backend/src/main/java/edu/kit/quak/files/model/File.java:
```diff
@@ -0,0 +1,89 @@
+    public void patch(File modified) {
+        super.patch(modified);
+        if (modified.createdOn != null && modified.createdOn.isAfter(createdOn))
+            this.createdOn = modified.createdOn;
```
Does it make sense to be able to change `createdOn`? Maybe this should be read only?

Henotu replied:
Fair point, one can no longer change the createdOn field

Jooshua commented on backend/src/main/java/edu/kit/quak/files/model/Project.java:
```diff
@@ -0,0 +1,74 @@
+        return content.add(element);
+    }
+
+    @Override
+    @JsonIgnore
+    public Collection<FileElement<?>> getContent() {
+        return new HashSet<>(content);
+    }
```
Maybe I am missing something, but in the API specification using GET `/project` returns a list of projects without contents, but using GET `/project/{pId}` the projects now have a content. Where is this behaviour defined?

Henotu replied:
The idea behind the containers sometimes displaying and sometimes not displaying their contents is to limit the amount of *unnecessary* information sent by the API. Similar to a folder in an OS, one only sees the entries that are directly located in a folder.

I added a clarifying description to the `/project/` endpoint, the other behaviour is mentioned in the top-level API-description

Jooshua reviewed (commented):
Looks good :) I just have a few minor questions.

Henotu committed:
Remove Gson from dependencies

Since Jackson is already used in spring, the tests now also use Jackson
to construct objects in JSON

Henotu committed:
Refactor Type-enum to dedicated classes

Henotu committed:
Refactor patch into FileElementSaver

Henotu committed:
Add deletion to FileElementSaver

Henotu committed:
Add filters to FileElementSaversRepository

Also added jdoc and removed some magic strings

Henotu committed:
Minor changes

Henotu committed:
Add contentType to File-metadata

Also fixed failing test

Henotu commented:
I now added the property `contentType` to the File-JSON.
During the meeting, I forgot that directories and files have to be POSTed with a valid `parent_id`.
Currently, the API's error messages aren't really helpful, which will get fixed in a future PR (see #11).

This PR is ready to be merged on my end.

Jooshua reviewed (approved):
Maybe we could remove the GSON dependence in another ticket. Looks good otherwise.
